### ISSUE TITLE
Temporarily remove multiple CNAMES from `registry-1.docker.io` DNS to work around a virtualbox DNS bug

### ISSUE THREAD

**Issue by sbuss** [docker/distribution#1292]
# tl;dr

Virtualbox corrupts DNS packets with more than one `CNAME` when `--natdnshostresolver1 on` is set (which is necessary for most company-internal private registries), which breaks pulls from `registry-1.docker.io`. Please remove the `CNAME` chain in the `registry-1.docker.io` DNS entry to work around this problem with virtualbox and resolve this problem for all docker users.

# Background

Virtualbox has a long-standing bug in their custom DNS resolver that corrupts DNS packets with multiple `CNAME`s. We, at [Counsyl](https://github.com/counsyl), opened a [ticket against virtualbox](https://www.virtualbox.org/ticket/14937) for the issue. This bug has apparently been in virtualbox for years but never reported. It is also present in the latest virtualbox. It's not clear if it will be fixed soon.

This DNS packet corruption results in an infinite DNS redirection loop when pulling from the public dockerhub when using a virtualbox `docker-machine` with the `--natdnshostresolver1 on` flag, as reported in https://github.com/boot2docker/boot2docker/issues/451

Note that this affects all virtualbox images, not just docker, but it has bitten a number of docker users using a virtualbox machine. And it can be solved with a simple workaround: remove the second `CNAME` entry in the `registry-1.docker.io` DNS response.

Please see my [gist](https://gist.github.com/sbuss/769da082fe987b9e0838) for `dig`s showing the buggy virtualbox behavior.

## Why not just disable `natdnshostresolver1`?

Counsyl uses a VPN and our private internal docker registry is only accessible/resolvable while connected the VPN. The only way to get virtualbox to resolve the registry's IP is to set `--natdnshostresolver1 on` on the `docker-machine`. This lets us access our own registry but often breaks access to the public registry. It's not yet clear why we are sometimes able to resolve the public registry, and sometimes not.

# Workaround

The DNS corruption only happens when a DNS response has multiple `CNAME` entries. It looks to be some kind of `CNAME` chain compression pointer bug. The bad codepath is not triggered when a DNS response only has a single `CNAME` in the response. This bug could be avoided for all docker users if the DNS response for the public registry managed to only use a single `CNAME` in the response.

# Thanks

Thanks to @hobbeswalsh for doing most of the hard work on figuring out the root cause, and @jamesob and @jbronn for answering various questions and helping our internal users work around the issue. CC @counsyl

**Comment by dmp42** [comment]
ping @bcwalrus @nwt @marcusmartins

(+ @docker/distribution-maintainers for the record)

**Comment by nwt** [comment]
@sbuss: Good news! We're down to one CNAME.
```
$ date
Tue Dec 22 09:17:24 PST 2015
$ dig +noall +answer registry-1.docker.io @8.8.8.8
registry-1.docker.io.	27	IN	CNAME	registry-origin.docker.io.
registry-origin.docker.io. 27	IN	A	52.0.101.18
registry-origin.docker.io. 27	IN	A	52.4.142.122
registry-origin.docker.io. 27	IN	A	54.208.179.150
```

**Comment by sbuss** [comment]
> Though this here is not really the right place for this report

Yeah :\ I wasn't sure where to report something like this.

> Good news! We're down to one CNAME.

@nwt Thanks, that was fast!

**Comment by hobbeswalsh** [comment]
awesome. thanks @sbuss and @nwt

**Comment by dmp42** [comment]
Cool. @sbuss issues with Hub are usually best served over http://support.docker.com - this here is the open-source registry codebase.

Either way, thanks to @nwt!

@sbuss Shall we close this ticket?

**Comment by sbuss** [comment]
Yup, all resolved. Closing! :+1:
